nasa/ow_simulator
# Oceanwater 700, 701 - activate comms action and camera power

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-697](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-697) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-700](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-700) |
| Jira Ticket 🎟️ | [OCEANWATER-701](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-701) |

## Summary of Changes
* Fixed light power consumption not being distributed across power nodes. 
* ActivateComms actions created. It will simply draw 28 W (settable in system.cfg) for 10 seconds.
* Power system node now listens to the /electrical_power topic and applies the power load it transmits evenly across all power cells. 
* Renamed `efficiency` to `motor_efficiency`.
* Camera controller, an always on device, now consumes power similar to lander lights and mechanical power. It is settable in `system.cfg` as `power_camera_controller`.
* Each camera being exposed draws a power of 2.0 W (see Caveats).

## Understanding the Wattage Values. 
Any power consumption applied to the battery, such as a single light at 100% intensity that consumes 0.3 W is divided by the number of still running nodes (4 if the `disconnect_battery_nodes` has not bee ticked). This means that a single light shows up in a cell's input power as 0.3 / 4 = 0.075 W if there are 4 power nodes operating. If one power node has failed the input power is 0.3 / 3  = 0.1 W.

## Test 1 - Activate Comms and Lights Power Fix
1. In `system.cfg`: increase`active_models` to at least 2, and change both `print_debug` and `inputs_print_debug` to true. 
2. Start any simulation world. 
3. The following will print
```
[ INFO:/power_system_node] M00 INPUT power: 7.500000.  volts: 17.675267.  tmp: 18.132230
[ INFO:/power_system_node] M01 INPUT power: 7.500000.  volts: 17.675267.  tmp: 18.132230
```
4. Tilt lander mast down so light project onto lander body `rosrun ow_lander tilt.py 1.2`
5. Shutdown one of the lights `rosrun ow_lander light_set_intensity.py left 0.0`
6. Verify output is now 
```
[ INFO:/power_system_node] M00 INPUT power: 6.300000.  volts: 17.662830.  tmp: 21.319773
[ INFO:/power_system_node] M01 INPUT power: 6.300000.  volts: 17.662830.  tmp: 21.319773
```
7. Shutdown the other light `rosrun ow_lander light_set_intensity.py right 0.0`
8. Verify the output
```
[ INFO:/power_system_node] M00 INPUT power: 5.100000.  volts: 17.644684.  tmp: 20.618757
[ INFO:/power_system_node] M01 INPUT power: 5.100000.  volts: 17.644684.  tmp: 20.618757
```
9. Turn on a single light `rosrun ow_lander light_set_intensity.py right 1.0` 
10. Increase the number of disconnected power cells by 1 by setting the value of `battery_nodes_to_disconnect` to 1, and ticking `disconnect_battery_nodes`. 
11. The output should now be
```
[ INFO:/power_system_node] M00 INPUT power: 6.833333.  volts: 17.657024.  tmp: 21.135746
[ INFO:/power_system_node] M01 INPUT power: 6.833333.  volts: 17.657024.  tmp: 21.135746
```
12. Power down the light `rosrun ow_lander light_set_intensity.py right 0.0` so that output is (the value is different because other power base lines are also being distributed across only 3 nodes now).
```
[ INFO:/power_system_node] M00 INPUT power: 5.233333.  volts: 17.627244.  tmp: 18.482299
[ INFO:/power_system_node] M01 INPUT power: 5.233333.  volts: 17.627244.  tmp: 18.482299
```
13. Run the ActivateComms action with default arguments to see that it draws 9.6 for 10 seconds (in simulator time). `rosrun ow_lander activate_comms.py`
15. The output while the action runs should be 
```
[ INFO:/power_system_node] M00 INPUT power: 8.433333.  volts: 17.609452.  tmp: 19.775166
[ INFO:/power_system_node] M01 INPUT power: 8.433333.  volts: 17.609452.  tmp: 19.775166
```
16. Run ActivateComms again with a different duration `rosrun ow_lander activate_comms.py --duration 5`. The output will be the same as above, but for only 5 seconds this time.

## Test 2 - Camera Power
1. In system.cfg: change both print_debug and inputs_print_debug to true. You can leave `active_models` at 2 or change it back to 1.
2. Start any simulation world.
3. You should see the output 
```
[ INFO:/power_system_node] M00 INPUT power: 7.500000.  volts: 17.650162.  tmp: 21.013723
```
* You will see more lines if `active_models` is still more than 1.  
4. Close the simulation
5. Change the value of `power_baseline_camera_controller` in system.cfg to `8.0`
6. Run the simulation, and you should see the following output
```
[ INFO:/power_system_node] M00 INPUT power: 9.100000.  volts: 17.667873.  tmp: 18.347697
```
Indicating that the extra camera controller power load of 6.4 has been distributed across all 4 power cells. 

## Caveat
The power draw from cameras being exposed has been implemented and is settable as `power_active_camera` in system.cfg; however, it appears our camera model does not actually take any simulation time to expose, no matter how high you set the exposure time. Meaning the value does not work as expected. The implementation is left for in case this camera exposure bug is addressed in the future. See [OW-1239](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1239)
